include/deal.II/matrix_free/evaluation_kernels.h

```diff
@@ -5748,17 +5761,53 @@ namespace internal
             evaluator.template hessians<1, true, false>(out, out);
           if (dim > 2)
             evaluator.template hessians<2, true, false>(out, out);
+        }
+      if (not dyadic_coefficients)
+        {
+          const unsigned int shift_coefficient =
+            inverse_coefficients.size() > dofs_per_component ?
+              dofs_per_component :
+              0;
+
+          for (unsigned int q = 0; q < dofs_per_component; ++q)
+            for (unsigned int d = 0; d < n_desired_components; ++d)
+              out_array[q + d * dofs_per_component] *=
+                inv_coefficient[q + d * shift_coefficient];
```

Comment:
I think it is a bit unfortunate to have this function work with indices in a different order than the dyadic coefficients you use below. In your new code, the tensor components are the fastest running index, whereas we here have the quadrature points running fastest. This is backwards compatibility, right? I would have preferred if we could keep the case that does not couple between components within one single loop over components (in order to reduce the data transfer from caches between the tensor product evaluator calls and the quadrature point operation), and then create a second copy of `evaluator.template hessians<...>(...)` that breaks out the loop as you need it for the case that couples between components. This creates more code, but we always do one or the other, so it won't affect the run time code access. If you do as @peterrum suggested, we will not even have a new function that gets instantiated but simply another path here.

Reply:
With https://github.com/dealii/dealii/pull/14850#issuecomment-1459658016, I was thinking about something like:
```cpp
n_comp_outer = dyadic_coefficients ? 1 : n_desired_components;
n_comp_inner = dyadic_coefficients ? n_desired_components : 1;

for (unsigned int d = 0; d < n_comp_outer; ++d)
  for (unsigned int di = 0; di < n_comp_inner; ++di)
    in_  = in  + dofs_per_component * di;
    out_ = out + dofs_per_component * di;
    evaluator.template hessians<0, true, false>(in_, out_);
    if (dim > 1)
      evaluator.template hessians<1, true, false>(out_, out_);
    if (dim > 2)
      evaluator.template hessians<2, true, false>(out_, out_);
      
  if(dyadic_coefficients)
    for(unsigned int q = 0; q < dofs_per_component; ++q)
      vmult(...);
  else
    for(unsigned int q = 0; q < dofs_per_component; ++q)
      out[q] *= inv_coefficient[q];
  
  for (unsigned int di = 0; di < n_comp_inner; ++di)
    in_  = in  + dofs_per_component * di;
    out_ = out + dofs_per_component * di;
    if (dim > 2)
      evaluator.template hessians<2, false, false>(out_, out_);
    if (dim > 1)
      evaluator.template hessians<1, false, false>(out_, out_);
    evaluator.template hessians<0, false, false>(out_, out_);
    
  in  += dofs_per_component;
  out += dofs_per_component;
```